src/vmspawn/vmspawn.c

```diff
@@ -426,20 +437,83 @@ static int cmdline_add_vsock(char ***cmdline, int vsock_fd) {
         assert(addr_len >= sizeof addr.vm);
         assert(addr.vm.svm_family == AF_VSOCK);
 
-        log_info("Using vsock-stream:%u:%u", (unsigned) VMADDR_CID_HOST, addr.vm.svm_port);
         r = strv_extendf(cmdline, "type=11,value=io.systemd.credential:vmm.notify_socket=vsock-stream:%u:%u", (unsigned) VMADDR_CID_HOST, addr.vm.svm_port);
         if (r < 0)
                 return r;
 
         return 0;
 }
 
+static int start_swtpm(sd_bus *bus, const char *scope, const char *swtpm, const char **ret_state_tempdir) {
+        _cleanup_(rm_rf_physical_and_freep) char *state = NULL;
+        _cleanup_strv_free_ char **cmdline = NULL, **cleanup = NULL;
+        _cleanup_free_ char *sock_path = NULL, *rm_path = NULL, *unit_name = NULL;
+        int r;
+
+        assert(bus);
+        assert(scope);
+        assert(swtpm);
+        assert(ret_state_tempdir);
+
+        r = mkdtemp_malloc("/tmp/vmspawn-swtpm-XXXXXX", &state);
+        if (r < 0)
+                return r;
+
+        sock_path = strjoin(state, "/sock");
+        if (!sock_path)
+                return log_oom();
+
+        cmdline = strv_new(swtpm, "socket", "--tpm2", "--tpmstate");
+        if (!cmdline)
+                return log_oom();
+
+        r = strv_extendf(&cmdline, "dir=%s", state);
+        if (r < 0)
+                return log_oom();
+
+        r = strv_extend_strv(&cmdline, STRV_MAKE("--ctrl", "type=unixio,fd=3"), /* filter_duplicates= */ false);
+        if (r < 0)
+                return log_oom();
+
+        r = find_executable("rm", &rm_path);
+        if (r < 0)
+                return log_error_errno(r, "Failed to find 'rm' binary: %m");
+
+        cleanup = strv_new(rm_path, "-rf", state);
+        if (!cleanup)
+                return log_oom();
+
+        unit_name = strjoin(scope, "-swtpm");
+        if (!unit_name)
+                return log_oom();
+
+        r = attach_command_to_socket_in_scope(bus, scope, unit_name, sock_path, SOCK_STREAM, cmdline, cleanup, NULL);
+        if (r < 0)
+                return r;
+
+        *ret_state_tempdir = TAKE_PTR(state);
+
+        return 0;
```

Comment:
hmm so if this only has a single parameter we could also make this a static template service, no? no need for the transient stuff.

i.e. a pair of regular template units vtpm@.socket + vptm@.service we'd just write to /usr/lib/systemd/system/ like any other unit.

The instance field would then be set to the full path of the state, using unit_name_path_escape() so that it can be referenced via %f in the template.

Reply:
How would this template the scope? From what I understand we need to use BindsTo in the unit otherwise they won't be stopped when vmspawn stops.

Reply:
They can to BindsTo=%i.scope?

Reply:
How would you tell virtiofsd what directory to host?